### PR TITLE
Transformer fix

### PR DESCRIPTION
**Pull request by Jyun1998**
## Proposed changes

Changes
- Changed order of layernorm and ffn according to tensorflow and pytorch implementations. (norm_first=False is default in pytroch as well as transformer)

- Added Encoder, Decoder with Dropout

Parts want to discuss

- Do you want to use Transformer with Dropout as main class?
- Do you like to add ffn as function or classes?
- Do you like to add norm_first parameter so that you could choose to do layer_norm first or later?
- Tensorflow do not do dropout with ffn while pytorch do. Which one should I refer to?

- I guess you did not add dropout since you would like to benchmark transformer without dropout for all jax, torch, tensorflow, mlx. I also had trouble adding dropout for jax. 

Testing
- All Encoder, Decoder (w and w/o dropout) are tested with mlx-exmaples/transformer_lm

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by awni** [comment]
> Do you want to use Transformer with Dropout as main class?

Let's just make dropout a kwarg parameter to the `Transformer*` which defaults to `0`. Don't need to implement seperate classes.

> Do you like to add ffn as function or classes?

Let's keep it the way it is implemented in `__call__`

> Do you like to add norm_first parameter so that you could choose to do layer_norm first or later?

I think we should have this. And maybe default it to `True` as that seems to be a more common setup?

> Tensorflow do not do dropout with ffn while pytorch do. Which one should I refer to?

Seems reasonable to follow PyTorch in that case. Add a comment about it when describing the `dropout` argument.

**Review by awni — CHANGES_REQUESTED** [review]
Thanks for the update! I responded to your questions. Please make the corresponding changes and then we can add this!

**Comment by awni** [comment]
Also do you mind adding a docstring to the `Transformer` class?

**Comment by awni** [comment]
@Jyun1998 are you still planning on updating this?

**Comment by Jyun1998** [comment]
> @Jyun1998 are you still planning on updating this?

Yes! I will finish it within this week with your comments :)

**Comment by Jyun1998** [comment]
Hi @awni ! Merry Chirstmas :) 

Added docstring, norm_first, and also activation to replace mx.maximum at ffn layer.

There are few points to further discuss.
- I have setted `norm_first=False,` which is common according to [pytorch_docstring](https://pytorch.org/docs/stable/_modules/torch/nn/modules/transformer.html#Transformer)

- by adding norm_first, codes got way extensive long. according to [pytorch_implementation](https://pytorch.org/docs/stable/_modules/torch/nn/modules/transformer.html#Transformer), they further break down encoder/decoer_layer with private function `_sa_block, _ff_block`. Do you want to keep it lastest commit or use `_sa_block, _ff_block`?


- default dropout ratio is 0 where Dropout __call__ is implemented with self._p_1 == 1, so Dropout(0) will do nothing.

File: python/mlx/nn/layers/transformer.py [hunk 1]
```diff
@@ -97,36 +99,71 @@ def create_additive_causal_mask(N: int, dtype: mx.Dtype = mx.float32):
...
-        y = self.ln2(x)
-        y = self.linear1(y)
-        y = mx.maximum(y, 0)
-        y = self.linear2(y)
-        x = x + y
+            y = self.linear1(y)
+            y = self.activation(y)
+            y = self.dropout2(y)
+            y = self.linear2(y)
+            y = self.ln2(x + y)
```
Comment (awni) [inline python/mlx/nn/layers/transformer.py:146]: I think the way you have it is long, but not bad because it is very clear. But, if you want the code to be a little shorter you could do something like this:

```
y = self.ln1(x) if self.norm_first else x
y = self.attention(y, y, y, mask)
y = x + self.dropout1(y)
y = y if self.norm_first else self.ln1(y) 

y = self.ln2(y) if self.norm_first else y
y = self.linear(y)
y = self.activation(y)
y = x + self.linear2(y)
y = y if self.norm_first else self.ln1(y) 
```

File: python/mlx/nn/layers/transformer.py [hunk 2]
```diff
@@ -189,29 +267,66 @@ def __call__(self, x, memory, x_mask, memory_mask):
...
+    Args:
+        dims (int): The number of expected features in the encoder/decoder inputs.
+        num_heads (int): The number of heads in the multi-head attention models.
+        num_encoder_layers (int): The number of sub-encoder-layers in the Transformer encoder.
+        num_decoder_layers (int): The number of sub-decoder-layers in the Transformer decoder.
+        mlp_dims (Optional[int]): The dimensionality of the feedforward network model in each Transformer layer.
+                                  Defaults to 4*dims if not provided.
+        dropout (float): The dropout value for Transformer encoder/decoder.
+        activation (Callable[[Any], Any]): the activation function of encoder/decoder intermediate layer
+        custom_encoder (Optional[Any]): A custom encoder to replace the standard Transformer encoder.
```
Comment (awni) [inline python/mlx/nn/layers/transformer.py:286]: Add defaults for these as well in the docstring.

**Review by awni — COMMENTED** [review]
Looks great, just a couple minor comments!

**Comment by awni** [comment]
> I have setted norm_first=False, which is common according to pytorch_docstring

Ok with me

> by adding norm_first, codes got way extensive long.

I left a comment that might work for shortening this. I don't think you need separate modules for these.

> default dropout ratio is 0 where Dropout call is implemented with self._p_1 == 1, so Dropout(0) will do nothing.

👍

**Comment by awni** [comment]
Also could you rebase? There is a merge conflict

**Comment by Jyun1998** [comment]
@awni Thanks for the review!

- I added defaults to doctring. 

- for the shortening the code, below will do the same but the code become quite not straight forward, so better left it as it is.

``` python
def __call__(self, x, mask):
    y = self.ln1(x) if self.norm_first else x
    y = self.attention(y, y, y, mask)
    y = self.dropout1(y)
    x = x + y  # Residual connection
    x = self.ln1(x) if not self.norm_first else x  # Apply LayerNorm if not norm_first

    y = self.ln2(x) if self.norm_first else x
    y = self.linear1(y)  # Corrected to linear1
    y = self.activation(y)
    y = self.dropout2(y)
    y = self.linear2(y)
    x = x + y  # Residual connection
    x = self.ln2(x) if not self.norm_first else x  # Apply LayerNorm if not norm_first

    return x
```

**Comment by Jyun1998** [comment]
Also rebased :)

**Review by awni — APPROVED** [review]
Looks great, thanks for adding!! I pushed a few changes to the doc strings and included stuff in the docs as well!

**Comment by Jyun1998** [comment]
> Looks great, thanks for adding!! I pushed a few changes to the doc strings and included stuff in the docs as well!

Thanks for the review :)
